libraries/abstractions/pkcs11/mbedtls/iot_pkcs11_mbedtls.c

```diff
@@ -219,43 +222,74 @@ typedef struct P11Session
     SemaphoreHandle_t xSignMutex;                /**< @brief Protects the signing key from being modified while in use. */
     mbedtls_pk_context xSignKey;                 /**< @brief Signing key.  Set during C_SignInit. */
     mbedtls_sha256_context xSHA256Context;       /**< @brief Context for in progress digest operation. */
-} P11Session_t, * P11SessionPtr_t;
+} P11Session_t;
 
+/*-----------------------------------------------------------*/
 
+/**
+ * @brief The global PKCS #11 module object.
+ * Entropy/randomness and object lists are shared across PKCS #11 sessions.
+ */
+static P11Struct_t xP11Context;
 
 /**
- * @brief Helper definitions.
+ * @brief Helper to check if the current session is initialized and valid.
  */
-#define PKCS11_MODULE_IS_INITIALIZED    ( ( xP11Context.xIsInitialized == CK_TRUE ) ? CK_TRUE : CK_FALSE )                                                                                          /**< Checks if PKCS #11 module is initialized. */
-#define PKCS11_SESSION_IS_OPEN( xSessionHandle )                         ( ( ( ( ( P11SessionPtr_t ) xSessionHandle )->xOpened ) == CK_TRUE ) ? CKR_OK : CKR_SESSION_CLOSED )                       /**< Checks if the current session is open */
-#define PKCS11_SESSION_IS_VALID( xSessionHandle )                        ( ( ( P11SessionPtr_t ) xSessionHandle != NULL ) ? PKCS11_SESSION_IS_OPEN( xSessionHandle ) : CKR_SESSION_HANDLE_INVALID ) /**< Checks if the current session is valid */
-#define PKCS11_SESSION_VALID_AND_MODULE_INITIALIZED( xSessionHandle )    ( PKCS11_MODULE_IS_INITIALIZED ? PKCS11_SESSION_IS_VALID( xSessionHandle ) : CKR_CRYPTOKI_NOT_INITIALIZED )                /**< Checks if the current session is valid and initialized. */
-/*-----------------------------------------------------------*/
-/*--------- See iot_pkcs11_pal.c for definitions ------------*/
+static CK_RV prvCheckValidSessionAndModule( const P11Session_t * pxSession )
+{
+    CK_RV xResult = CKR_OK;
 
+    /* An integer constant expression with the value 0 or 1 of either signedness
```

Comment:
I think we can make the explanation a little better here. What do you think about the below:

```
MISRA Rule 10.5 - Cannot cast from unsigned to signed.

The rule 10.5 is violated because type of xP11Context.xIsInitialized is <whatever> which is signed and the underlying type of CK_BBOOL is unsigned char.

It is justified to suppress this violation because the expression is casting CK_FALSE  which is constant and can have only 0/1 values (Or whatever the explanation is).
```

Basically briefly stating what the rule is, what is causing the violation and why is it justified.

Reply:
Updated this description